File: lib/bridge/bls.ffi.dart
import 'dart:typed_data';

import 'package:agent_dart/bridge/bls.base.dart';
import 'ffi/ffi.dart';

class FFIBls implements BaseBLS {
  late bool _isInit = false;

  @override
  Future<bool> blsInit() async {
    _isInit = await AgentDartFFI.impl.blsInit();
    return _isInit;
  }

  @override
  Future<bool> blsVerify(
    Uint8List pk,
    Uint8List sig,
    Uint8List msg,
  ) {
    return AgentDartFFI.impl.blsVerify(
      req: BLSVerifyReq(signature: sig, message: msg, publicKey: pk),
    );
  }

  @override
  bool get isInit => _isInit;
}

BaseBLS createBLS() => FFIBls();

File: packages/agent_dart_ffi/native/agent_dart/src/bls_ffi.rs
use crate::bls::bls12381::bls::{core_verify, init, BLS_OK};
use crate::types::BLSVerifyReq;

pub struct BlsFFI {}

impl BlsFFI {
    pub fn bls_init() -> bool {
        init() == BLS_OK
    }

    pub fn bls_verify(req: BLSVerifyReq) -> bool {
        let verify = core_verify(
            req.signature.as_slice(),
            req.message.as_slice(),
            req.public_key.as_slice(),
        );
        verify == BLS_OK
    }
}
